File: 543_diameter_of_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int[] res = new int[] {0};
        calPathLeng(root, res);
        return res[0];
    }
    
    private int calPathLeng(TreeNode node, int[] res) {
        if (node == null) return 0;
        int leftLen = calPathLeng(node.left, res);
        System.out.println("left len: " + leftLen);
        int rightLen = calPathLeng(node.right, res);
        System.out.println("right len: " + rightLen);
        int sum = leftLen + rightLen;
        if (res[0] < sum) {
            res[0] = sum;
        }
        return(leftLen > rightLen ? leftLen + 1 : rightLen + 1); 
    }
}

File: python/543_diameterof_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        # calculate the length of left node and right node
        # thinking the depth starting from the bottom of a leaf node
        self.ans = 0
        def depth(node):
            if not node: return 0
            left = depth(node.left)
            right = depth(node.right)
            # path
            self.ans = max(self.ans, left + right)
            # depth
            return max(left, right) + 1
        depth(root)
        return self.ans
